wallet/wallet.go

```diff
@@ -410,6 +453,128 @@ func (sw *SingleAddressWallet) UnconfirmedTransactions() ([]Transaction, error)
 	return annotated, nil
 }
 
+// Redistribute returns a transaction that redistributes money in the wallet by
+// selecting a minimal set of inputs to cover the creation of the requested
+// outputs. It also returns a list of output IDs that need to be signed.
+func (sw *SingleAddressWallet) Redistribute(outputs int, amount, feePerByte types.Currency) (txns []types.Transaction, toSign []types.Hash256, err error) {
+	// fetch outputs from the store
+	utxos, err := sw.store.UnspentSiacoinElements()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// fetch outputs currently in the pool
+	inPool := make(map[types.Hash256]bool)
+	for _, txn := range sw.cm.PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			inPool[types.Hash256(sci.ParentID)] = true
+		}
+	}
+
+	// grab current height
+	state := sw.cm.TipState()
+	bh := state.Index.Height
+
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	// adjust the number of desired outputs for any output we encounter that is
+	// unused, matured and has the same value
+	usable := utxos[:0]
+	for _, sce := range utxos {
+		inUse := time.Now().After(sw.locked[sce.ID]) || inPool[sce.ID]
+		matured := bh >= sce.MaturityHeight
+		sameValue := sce.SiacoinOutput.Value.Equals(amount)
+
+		// adjust number of desired outputs
+		if !inUse && matured && sameValue {
+			outputs--
+		}
+
+		// collect usable outputs for defragging
+		if !inUse && matured && !sameValue {
+			usable = append(usable, sce)
+		}
+	}
+
+	// return early if we don't have to defrag at all
+	if outputs <= 0 {
+		return nil, nil, nil
+	}
+
+	// in case of an error we need to free all inputs
+	defer func() {
+		if err != nil {
+			for _, id := range toSign {
+				delete(sw.locked, id)
+			}
+		}
+	}()
+
+	// desc sort
+	sort.Slice(usable, func(i, j int) bool {
+		return usable[i].SiacoinOutput.Value.Cmp(utxos[j].SiacoinOutput.Value) > 0
+	})
+
+	// prepare defrag transactions
+	for outputs > 0 {
+		var txn types.Transaction
+		for i := 0; i < outputs && i < RedistributeBatchSize; i++ {
+			txn.SiacoinOutputs = append(txn.SiacoinOutputs, types.SiacoinOutput{
+				Value:   amount,
+				Address: sw.addr,
+			})
+		}
+		outputs -= len(txn.SiacoinOutputs)
+
+		// estimate the fees
+		outputFees := feePerByte.Mul64(state.TransactionWeight(txn))
```

Comment:
A while ago we had the notion that transactions should have an abstract "weight" instead of simply their encoded size in bytes. For example, we might want to consider a transaction with a lot of signatures "heavier" than a transaction with a lot of arbitrary data, because validating such a transaction imposes higher resource costs.

This was eventually abandoned, but we still define these as methods on `consensus.State` because they *are* directly relevant to consensus rules. If we someday decide to change how the weight is computed, we can do that by updating the single `TransactionWeight` method, instead of finding and fixing every piece of code that manually calculates a transaction's encoded size.

Reply:
The idea is that anything tied to a consensus rule should be tied to a `consensus.State`, because `consensus.State` knows what hardfork it's on. If we wrote `TransactionWeight` the straightforward way, it would just take a `types.Transaction` and no other arguments. Later, if there was a hardfork that changed the behavior, we would need to modify the function to take a `consensus.State` (or at least a `height uint64`). And then we would need to update the callsites too, because those callsites probably weren't written with the assumption that they would need the current height... and then update the callers of *those* callers, and so on.

(This is exactly what happened in `siad` with the `(Transaction).SigHash` method. We *still* see code where that method is called with a fake height, e.g. `types.FoundationHardforkHeight+1`, because the surrounding code wasn't written with the assumption that it would need to know the height. That's the situation that this API is aiming to prevent.)

Reply:
Thanks for clarifying that 👍 makes sense now.